packages/bvaughn-architecture-demo/src/suspense/SearchCache.ts

```diff
@@ -0,0 +1,180 @@
+} = createGenericCache2<
+  ReplayClientInterface,
+  [query: string, includeNodeModules: boolean, limit?: number],
+  StreamingSourceSearchResults
+>(
+  async (
+    client: ReplayClientInterface,
+    query: string,
+    includeNodeModules: boolean,
+    limit: number = MAX_SEARCH_RESULTS_TO_DISPLAY
+  ) => {
+    const subscribers: Set<Subscriber> = new Set();
+
+    const orderedResults: SourceSearchResult[] = [];
+
+    const result: StreamingSourceSearchResults = {
+      complete: false,
+      fetchedCount: 0,
+      orderedResults,
+      subscribe: (subscriber: Subscriber) => {
+        subscribers.add(subscriber);
+        return () => {
+          subscribers.delete(subscriber);
+        };
+      },
+      totalCount: 0,
+    };
+
+    if (sourceIdsWithNodeModules === null) {
+      await initializeSourceIds(client);
+    }
+
+    const sourceIds = includeNodeModules ? sourceIdsWithNodeModules! : sourceIdsWithoutNodeModules!;
+
+    const notifySubscribers = () => {
+      subscribers.forEach(subscriber => subscriber());
+    };
+
+    let currentResultLocation: SourceSearchResultLocation | null = null;
+
+    client
+      .searchSources({ query, sourceIds }, (matches: SearchSourceContentsMatch[]) => {
+        result.totalCount += matches.length;
+
+        for (let index = 0; index < matches.length; index++) {
+          if (result.fetchedCount >= limit) {
+            break;
+          }
+
+          result.fetchedCount++;
+
+          const match = matches[index];
+
+          if (
+            currentResultLocation === null ||
+            currentResultLocation.location.sourceId !== match.location.sourceId
+          ) {
+            currentResultLocation = { location: match.location, matchCount: 0, type: "location" };
+
+            orderedResults.push(currentResultLocation);
+          }
+
+          currentResultLocation.matchCount++;
+
+          orderedResults.push({ match, type: "match" });
+        }
+
+        notifySubscribers();
+      })
+      .then(() => {
+        result.complete = true;
+
+        notifySubscribers();
+      });
+
+    return result;
+  },
+  (query: string, includeNodeModules: boolean, limit?: number) =>
+    `${includeNodeModules}:${limit || "-"}:${query}`
+);
+
+async function initializeSourceIds(client: ReplayClientInterface) {
+  sourceIdsWithNodeModules = [];
+  sourceIdsWithoutNodeModules = [];
+
+  const sources = await getSourcesAsync(client);
+
+  // Insert sources in order so that original sources are first.
+  const compareSources = (a: SourceId, b: SourceId) => {
+    const aIsOriginal = client.isOriginalSource(a);
+    const bIsOriginal = client.isOriginalSource(b);
+    if (aIsOriginal === bIsOriginal) {
+      return 0;
+    } else if (aIsOriginal) {
+      return -1;
+    } else {
+      return 1;
+    }
+  };
+
+  sources.forEach(source => {
+    const sourceId = source.sourceId;
+
+    if (client.isPrettyPrintedSource(sourceId)) {
+      return;
```

Comment:
This means we're using minified sources instead of their pretty-printed versions but I think we should do it the other way around.

Reply:
Hmm, this is what the old implementation did:
```js
async function initializeSourceIds(client: ReplayClientInterface) {
  sourceIdsWithNodeModules = [];
  sourceIdsWithoutNodeModules = [];

  const sources = await getSourcesAsync(client);

  // Insert sources in order so that original sources are first.
  const compareSources = (a: SourceId, b: SourceId) => {
    const aIsOriginal = client.isOriginalSource(a);
    const bIsOriginal = client.isOriginalSource(b);
    if (aIsOriginal === bIsOriginal) {
      return 0;
    } else if (aIsOriginal) {
      return -1;
    } else {
      return 1;
    }
  };

  sources.forEach(source => {
    const sourceId = source.sourceId;

    if (client.isPrettyPrintedSource(sourceId)) {
      return;
    }
```

Reply:
I agree with you that it seems like using the pretty-printed one would be better if there are both. Will we _always have_ a pretty-printed version?

Reply:
No, we don't always have a pretty-printed version, only if the file is minified (i.e. if it contains a line with more than 500 characters). So we'd have to check if there is a pretty-printed version for a non-pretty-printed file before skipping it.

Reply:
How would we do that?

Reply:
Check if there is a pretty-printed source that has the current source as its "generated" source ("generated" according to the backend, in the case of pretty-printing the backend declares the minified source as the generated one, even though in reality it's the other way around).

Reply:
I filed FE-983 for this follow up.